src/collection/shared/generic_collector.py

```diff
@@ -0,0 +1,174 @@
+            LOG.error(f"Transaction `{self.tx_body.transaction.signatures[0]}`"
+                      f" does not contain any relevant public keys.")
+
+        return relevant_sources
+
+    @property
+    def signature(self) -> Signature:
+        """ Returns transaction signature. """
+        return self.tx_body.transaction.signatures[0]
+
+
+class GenericSolanaConnector(ABC):
+    """
+    Abstract class with methods, that every Solana connector should implement.
+    """
+
+    def __init__(self):
+        self._get_rpc_url()
+        self.solana_client: solana.rpc.api.Client = solana.rpc.api.Client(self.authenticated_rpc_url)
+
+        self.assignment: List[int] = list()
+
+        # attribute for storing transactions before assigning to db.
+        self.rel_transactions: List[SolanaTransaction] = list()
+
+        # rate limiter
+        self.rate_limit: int = int(os.getenv("RATE_LIMIT", ""))
+        self._call_timestamps = []
+
+        LOG.info(f"{self.__class__.__name__} is all set to collect.")
+
+    def _get_rpc_url(self) -> None:
+        rpc_url = os.getenv("RPC_URL")
+        if not rpc_url:
+            LOG.error("RPC url was not found in environment variables.")
+        self.authenticated_rpc_url = rpc_url
+
+    def run(self):
+        """
+        Main method to process all necessary operations to collect and write data.
+        """
+        k = 0
+        start_time = time.time()
+        while not self._collection_completed:
+            if k % 10000 == 0:
+                LOG.info(f"Iterations completed: {k}, in {time.time() - start_time:.1f} seconds.")
+            self._get_assignment()
+            self._get_data()
+            self._write_tx_data()
+            k += 1
+        LOG.info(f"Collection completed for {self.__class__.__name__}.")
+
+    def _fetch_block(self, block_number: int) -> Tuple[UiConfirmedBlock, int]:
+        """
+        Use solana client to fetch block with provided number.
+        """
+        self._rate_limit_calls()
+        # Fetch block data.
+        try:
+            block = self.solana_client.get_block(
+                slot=block_number,
+                encoding='jsonParsed',
+                max_supported_transaction_version=0,
+            )
+        except SolanaRpcException as e:
+            LOG.error(f"SolanaRpcException: {e}")
+            time.sleep(1)
+            return self._fetch_block(block_number)
+
+        return block.value, block_number
+
+    def _rate_limit_calls(self) -> None:
+        """
+        This method implements a rate-limiting mechanism to control the frequency of API calls.
+        It ensures that the number of API calls does not exceed a predefined limit per second.
+        If the limit is reached, it will pause the execution momentarily before allowing further API calls.
+        """
+        # Check if the number of calls made in the last second exceeds the maximum limit.
+        while self.rate_limit <= len(self._call_timestamps):
```

Comment:
https://docs.python.org/3/library/collections.html#collections.deque would perhaps lead to a more elegant solution. Just an idea, there's no need to change anything.

Reply:
tried, but my sense of elegance is quite limited. please take a look.

Reply:
Lgtm :+1: